### ISSUE TITLE
`structure.unused_cte` misses usage in nested CTEs

### ISSUE THREAD

**Issue by elyobo** [sqlfluff/sqlfluff#5561]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

`structure.unused_cte` detects unused CTEs, but does not detect usage of CTEs within nested CTEs (CTEs defined by a `WITH` statement inside another CTE), causing it to report used CTEs as unused.

### Expected Behaviour

`structure.unused_cte` should detect all uses of a CTE even in nested CTEs.

### Observed Behaviour

See reproduction example below.

### How to reproduce

The following example uses `example_cte` inside `nested_cte` which is defined by a  `WITH` statement inside `container_cte`.

```sql
WITH
  example_cte AS (
    SELECT 1,
  ),

  container_cte AS (
    WITH
      nested_cte AS (
        SELECT *,
        FROM example_cte
      )

    SELECT *,
    FROM nested_cte
  )

SELECT *,
FROM container_cte
```

The `.sqlfluff` in use is described in the configuration section below.

linting the file reports

```
poetry run sqlfluff lint models/example.sql
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [models/example.sql] FAIL
L:   2 | P:   3 | ST03 | Query defines CTE "example_cte" but does not use it.
                       | [structure.unused_cte]
All Finished 📜 🎉!
```

### Dialect

BigQuery

### Version

```
✗ poetry run sqlfluff --version
sqlfluff, version 2.3.5
```

```
✗ python --version
Python 3.11.0
```

### Configuration
```
[sqlfluff]
verbose = 0
nocolor = false
dialect = bigquery
templater = dbt
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = true
encoding = autodetect
max_line_length = 100
# Comma separated list of file extensions to lint.
# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

# Disable some rules
# See https://docs.sqlfluff.com/en/stable/rules.html
# LT07/L018/layout.cte_bracket: Align WITH brackets, incompatible with indented_ctes
# AL07/L031/aliasing.forbid: Allow aliasing in FROM/JOIN
# ST07/L032/structure.using: Allow USING statement
# ST06/L034/structure.column_order: Disable wildcard over specific select (L034) - it
#   causes some unwanted behaviour, specifically complaining about select * in
#   subqueries
exclude_rules = layout.cte_bracket, aliasing.forbid, structure.using, structure.column_order

[sqlfluff:indentation]
indented_joins = true
indented_ctes = true
tab_space_size = 2
indented_on_contents = false

[sqlfluff:templater]
unwrap_wrapped_queries = true

[sqlfluff:templater:jinja]
apply_dbt_builtins = true

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = "{% macro ref(model_ref) %}{{model_ref}}{% endmacro %}"
dbt_source = "{% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}"
dbt_config = "{% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}"
dbt_var = "{% macro var(variable, default='') %}item{% endmacro %}"
dbt_is_incremental = "{% macro is_incremental() %}true{% endmacro %}"
load_macros_from_path = "macros"

# Some rules can be configured directly from the config common to other rules.
# NOTE(@elyobo): override non default config only please.
# For defaults see https://docs.sqlfluff.com/en/stable/configuration.html#defaultconfig

[sqlfluff:rules:capitalisation.keywords]
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CP01
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.identifiers]
# Identifier capitalisation
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CP02
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
# Function casing
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CP03
extended_capitalisation_policy = upper

[sqlfluff:rules:capitalisation.literals]
# Identifier capitalisation
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CP04
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.types]
# Identifier capitalisation
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CP05
extended_capitalisation_policy = upper

[sqlfluff:rules:convention.select_trailing_comma]
# Trailing commas in SELECT
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CV03
select_clause_trailing_comma = require
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by sdepablos** [comment]
I was able to replicate the issue, with BigQuery and version 2.3.1. Version 2.3.0 works fine. 

I've also tried a `git bisect` between 2.3.0 and 2.3.1, and this is the first commit https://github.com/sqlfluff/sqlfluff/commit/68ef48051d09cf543c5b439f04a5850f49731e7a with the issue, corresponding to the PR #5104, most probably this part that excludes inner CTEs.

```python
        # Work through all the references in the file, checking off CTES as the
        # are referenced. We don't recurse inside inner WITH statements.
        for reference in context.segment.recursive_crawl(
            "table_reference", no_recursive_seg_type="with_compound_statement"
        ):
            remaining_ctes.pop(reference.raw.upper(), None)
```

**Comment by sdepablos** [comment]
I'm trying version 3.0.3 and the problem persists :(

**Comment by sdepablos** [comment]
@will-warner AFAIK nobody is working on this, so my recommendation is adding ST03 to the sqlfluffignore for now.

**Comment by arturpeedimaa** [comment]
I upgraded from v2.1.1 to the latest version and also stumbled onto this issued.
I'm using macros within dbt which compile sql with nested cte-s (~50 models), seems a bit annoying that it's not able to handle this case anymore and would be extremely happy if it was addressed..

**Comment by sdepablos** [comment]
@will-warner @elyobo the pull request #6017 solves the issue, now we only need for it to be merged
